### ISSUE TITLE
static layer sometimes doesn't use map to create the global costmap

### ISSUE THREAD

**Issue by bfalacerda** [ros-planning/navigation#187]
Sometimes, the static layer doesn't create the global costmap properly, creating only an empty costmap. 

To exemplify, this is the map being published:

![screenshot from 2014-04-28 17 44 24](https://cloud.githubusercontent.com/assets/4320588/2819963/9197824c-cef4-11e3-8d6d-e72ebc4dbc30.png)

This is the global costmap when the static layer doesn't load the map properly

![screenshot from 2014-04-28 17 44 51](https://cloud.githubusercontent.com/assets/4320588/2819966/a395821e-cef4-11e3-9fae-f1114de818b9.png)

If you try to relaunch, it loads the map properly, and you have the following global costmap:

![screenshot from 2014-04-28 17 45 09](https://cloud.githubusercontent.com/assets/4320588/2819979/da74736c-cef4-11e3-9de6-7dcce3232899.png)

This problem seems to only appear the first time you try to launch navigation, after that it looks like it always creates the global costmap properly

**Comment by mikeferguson** [comment]
I've also seen this, not sure the cause though.

**Comment by stonier** [comment]
Also seeing I think the same symptom. 
- My global costmap loaded up only seeing the local costmap drawn on it (probably the same as above where there is a single blob next to the robot). 
- After driving around for a bit, suddenly the whole global costmap appeared
- Driving around a bit more, and it became obvious that it was drawing the local costmap on top of the global costmap.

First picture below is just the global costmap after driving around a bit. Note the parts of the walls left and above of the robot where the walls are 'broken'. This is where the local costmap is currently overlaying - the second picture shows the actual local costmap.

![global_costmap](https://cloud.githubusercontent.com/assets/687378/3163206/60b91f0e-eb3d-11e3-8820-13648dadbdfa.png)

![local_costmap](https://cloud.githubusercontent.com/assets/687378/3163209/6ea66a90-eb3d-11e3-9bf7-9a505c9b26b7.png)

**Comment by stonier** [comment]
So two problems above:
- Why is it not getting the global costmap till after it has been driving around for some time.
- Is it really supposed to paint the local costmap onto the global costmap?

Some other odd things:
- `rostopic hz /move_base/global_costmap/costmap` hangs (i.e. says its not getting new messages...)
- but if you frequently `rostopic echo ...`, CTRL-C, `rostopic echo` the timestamp is changing
- the header sequence number is however continuously zero.

Is that correct behaviour?

**Comment by mikeferguson** [comment]
> rostopic hz /move_base/global_costmap/costmap hangs (i.e. says its not getting new messages...)
> but if you frequently rostopic echo ..., CTRL-C, rostopic echo the timestamp is changing

This is expected if your costmap publish frequency is 0.0 as the only thing that causes a message to be published is a new subscriber. The publisher has a callback for new subscriptions (https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/src/costmap_2d_publisher.cpp#L76) which will update the grid and republish.

**Comment by stonier** [comment]
> This is expected if your costmap publish frequency is 0.0 as the only thing that causes a message to be published is a new subscriber. The publisher has a callback for new subscriptions (https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/src/costmap_2d_publisher.cpp#L76) which will update the grid and republish.

Got it, thanks Fergs.

**Comment by mikeferguson** [comment]
I started to look at this for a moment today (cause it has been a problem for us too), but I got distracted before I got to the bottom of it.

@stonier you said "Why is it not getting the global costmap till after it has been driving around for some time." -- that's not exactly what I've seen. In general, the first time I start the navigation stack I often don't see any map in the static costmap, ever. Restarting navigation gets me a map. Do you actually see the map occur at some point later? If so, about how long after you start up, or how far do you drive?

**Comment by stonier** [comment]
Some more testing this morning reveals:

When the static layer of the global costmap initially fails:
- global costmap gets to fully publish when the 'Clearing costmap to unstuck robot.' event occurs.
  - this resets grid size and width and [allows](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/src/costmap_2d_publisher.cpp#L115) the periodic republisher to do a full publication on the next loop
  - at this point the costmap is usually finally initialised with the underlying `/map`.

It looks like the logic for periodically publishing the global costmap isn't working. It will initially post an empty static layer and when it finally receives the underlying map the full publish is skipped because resolution and size hasn't changed (even if the underlying map data has).

An inefficient workaround for now seems to be to set the global costmap parameter `always_send_full_costmap: true`.

**Comment by stonier** [comment]
>  I often don't see any map in the static costmap, ever. Restarting navigation gets me a map. Do you actually see the map occur at some point later? If so, about how long after you start up, or how far do you drive?

Yeah, I had that as well sometimes. Drive around for ages and nothing happened. I think I hit on the trigger this morning though - the unstuck event which clears everything and forces a full republish (haven't managed to trace this through the code yet though).

**Comment by stonier** [comment]
One thing doesn't make sense still. My first new subscription to the global costmap publisher occurs long after the `static_layer` plugin has retrieved the map.

**Comment by mikeferguson** [comment]
> global costmap gets to fully publish when the 'Clearing costmap to unstuck robot.' event occurs.

This is a very good data point, I hadn't yet come up with what a trigger might be.

**Comment by KaijenHsiao** [comment]
I just started noticing the same problem a few days ago and looking into it today, what a coincidence. :)  Even with always_send_full_costmap set to true I still have the problem sometimes.

**Comment by stonier** [comment]
> always_send_full_costmap set to true I still have the problem sometimes

I wasn't 100% on that one, good to know!

**Comment by stonier** [comment]
- static layer has good information from the map, even when the global costmap isn't getting published
- when there is an issue, the global costmap is quite often printing a zero'd map

So I decided to look for places where the layered costmap is getting reset.

In [updateMap()](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/src/layered_costmap.cpp#L78) it resets the map before updating it from the static layer plugin (and others as well). In particular, this code:

```
  costmap_.resetMap(x0, y0, xn, yn);

  {
    boost::unique_lock < boost::shared_mutex > lock(*(costmap_.getLock()));
    for (vector<boost::shared_ptr<Layer> >::iterator plugin = plugins_.begin(); plugin != plugins_.end();
        ++plugin)
    {
      (*plugin)->updateCosts(costmap_, x0, y0, xn, yn);
    }
  }
```

It looks to me like there's an opportunity for the [publisher](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/src/costmap_2d_publisher.cpp#L76) to squeeze in a publication of a reset map (the costmap lock is also used inside resetMap()) inbetween the reset call and the locked scope where it updates from plugins.

**Comment by stonier** [comment]
The above is probably still a valid concern, although miniscule in its chance of occuring, but I couldn't prove the assertion while testing.  It is something else and happening far more frequently than the above situation would suggest.

**Comment by stonier** [comment]
A better lead:
- The obstacle layer plugin is wiping out the global costmap on the first update.
- When it works, the obstacle layer updates via updateWithMax
- When it fails, the obstacle layer is updating viz updateWithOverwrite

See the obstacle layer update code [here](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/plugins/obstacle_layer.cpp#L382).

The variable which triggers this is getting ambiguously(?) set via dynamic reconfigure?

Example of logging from failure (the max value is simply the maximum value in the costmap):

![failed_to_update](https://cloud.githubusercontent.com/assets/687378/3171191/0a850b3a-ebc3-11e3-8107-903d48506cb9.png)

Example of logging from success:

![updated](https://cloud.githubusercontent.com/assets/687378/3171188/050b1a28-ebc3-11e3-905b-434b828949a4.png)

**Comment by stonier** [comment]
- `move_base` is using `VoxelLayer` for the obstacle layer plugin for both global and local costmaps.
-  `VoxelLayer` is a child of `ObstacleLayer`
- VoxelLayer has its own dynamic reconfigure server different to obstacle layer's.
- VoxelLayer's DR server does not serve the variable `combination_method`, ObstacleLayer does.
- VoxelLayer updates via [ObstacleLayer::updateCosts](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/plugins/obstacle_layer.cpp#L373), it doesn't have it's own implementation of this function.
- In updateCosts it tries to access the uninitialised combination_method variable.

In a nutshell, [VoxelPlugin.cfg](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/cfg/VoxelPlugin.cfg) either needs to pull in [ObstaclePlugin.cfg](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/cfg/ObstaclePlugin.cfg) or set the `combination_method` variable itself.

**Comment by stonier** [comment]
Bugs are a great way of _really_ learning how things work sometimes! :neckbeard:

**Comment by mikeferguson** [comment]
@stonier so does one of the updates works but the other does not? If so, which version does not work?

**Comment by stonier** [comment]
@mikeferguson are you talking about the `updateCosts()` functions? There is only one of them as VoxelLayer doesn't implement it.

**ObstacleLayer**: probably works fine, but I haven't tested move base with instances of this plugin.
- defines a protected variable [combination_method](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/include/costmap_2d/obstacle_layer.h#L178)
- uses [ObstaclePlugin.cfg](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/cfg/ObstaclePlugin.cfg) to configure and [set](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/plugins/obstacle_layer.cpp#L198) `combination_method`.
- runs [ObstacleLayer:updateCosts()](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/plugins/obstacle_layer.cpp#L373) which makes use of `combination_method`.

**VoxelLayer**: this is the one that causes the above problems when used with move_base.
- inherits variable `combination_method` from parent `ObstacleLayer`.
- uses [VoxelPlugin.cfg](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/cfg/VoxelPlugin.cfg) which does not get used to set `combination_method` as above.
- runs [ObstacleLayer:updateCosts()](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/plugins/obstacle_layer.cpp#L373) which uses ambiguous `combination_method`.

**Comment by stonier** [comment]
Just to test, I added the following line to the end of initialisation in [voxel_layer.cpp](https://github.com/ros-planning/navigation/blob/hydro-devel/costmap_2d/plugins/voxel_layer.cpp#L28)

```
  combination_method_ = 1;
```

and ran a stage simulation 20 times. 0% failure. Previously it was between 25-50% fail.

**Comment by DLu** [comment]
+1

My apologies for not being active in the discussion, as I was on vacation. But thanks for the bug fix!

**Comment by spmaniato** [comment]
Apologies for resurrecting a 2.5 year old issue, but it seems like the param `combination_method` is not exposed on the parameter server nor is it initialized (`onInitialize`). It's only exposed as a dynamic parameter and initialized in the dynamic reconfigure callback (c.f. #202). Was this a conscious design choice? If so, why? Thank you 😁 

PS1. I'm on Indigo, but I don't see a fix in other recent distros either.
PS2. Looking at some of the commits that reference this issue, it seems like people are initializing it in their forks of the nav stack. Would it make sense to open a PR or is it a breaking change?
